hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAssociateToAccountHandler.java

```diff
@@ -42,22 +67,223 @@ public TokenAssociateToAccountHandler() {
+            @NonNull final ReadableTokenStore tokenStore,
+            @NonNull final WritableTokenRelationStore tokenRelStore) {
+        requireNonNull(config);
+
+        // Check that the system hasn't reached its limit of token associations
+        var maxTokenRels = config.getValue("maxNumTokenRels", Long.class);
+        validateTrue(
+                isTotalNumTokenRelsWithinMax(tokenIds.size(), tokenRelStore, maxTokenRels),
+                MAX_ENTITIES_IN_PRICE_REGIME_HAVE_BEEN_CREATED);
+
+        // Check that the account exists
+        final var account = accountStore.get(accountId).orElse(null);
+        validateTrue(account != null, INVALID_ACCOUNT_ID);
+
+        // Check that the given tokens exist and are usable
+        final var tokens = new ArrayList<Token>();
+        for (final TokenID tokenId : tokenIds) {
+            final var token = tokenStore.getToken(tokenId);
+            validateTrue(token != null, INVALID_TOKEN_ID);
+            validateFalse(token.deleted(), TOKEN_WAS_DELETED);
+            validateFalse(token.paused(), TOKEN_IS_PAUSED);
+            tokens.add(token);
+        }
+
+        // Check that the total number of old and new token IDs wouldn't be bigger than
+        // the max number of token associations allowed per account (if the rel limit is enabled)
+        validateTrue(
+                maxAccountAssociationsAllowTokenRels(config, account, tokenIds), TOKENS_PER_ACCOUNT_LIMIT_EXCEEDED);
+
+        // Check that a token rel doesn't already exist for each new token ID
+        for (final TokenID tokenId : tokenIds) {
+            final var existingTokenRel = tokenRelStore
+                    .get(tokenId.tokenNum(), accountId.accountNumOrThrow())
+                    .orElse(null);
+            validateTrue(existingTokenRel == null, TOKEN_ALREADY_ASSOCIATED_TO_ACCOUNT);
+        }
+
+        return new Validated(account, tokens);
+    }
+
+    private boolean isTotalNumTokenRelsWithinMax(
+            final int numNewTokenRels, WritableTokenRelationStore tokenRelStore, long maxNumTokenRels) {
+        return tokenRelStore.sizeOfState() + numNewTokenRels <= maxNumTokenRels;
+    }
+
+    /**
+     * Method that checks if the number of token associations for the given account is within the
+     * allowable limit set by the config (if the limit is enabled)
+     *
+     * @return true if tokenAssociationsLimited is false or if the number of token associations is
+     * within the allowed maxTokensPerAccount
+     */
+    private boolean maxAccountAssociationsAllowTokenRels(
+            @NonNull Configuration config, @NonNull Account account, @NonNull List<TokenID> tokenIds) {
+        final var numAssociations = requireNonNull(account).numberAssociations();
+        final var tokenAssociationsLimited = config.getValue("areTokenAssociationsLimited", Boolean.class);
+        final var maxTokensPerAccount = config.getValue("maxTokensPerAccount", Long.class);
```

Comment:
same for these configs, can be added to `TokenServiceConfig`

Reply:
Sure, I'll add these after #6582 is merged

Reply:
Just an info, Merged the PR.

Reply:
Done